services/elasticache/pkg/resource/replication_group/custom_update_api.go

```diff
@@ -0,0 +1,388 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package replication_group
+
+import (
+	"context"
+	"fmt"
+	ackcompare "github.com/aws/aws-controllers-k8s/pkg/compare"
+	svcapitypes "github.com/aws/aws-controllers-k8s/services/elasticache/apis/v1alpha1"
+	svcsdk "github.com/aws/aws-sdk-go/service/elasticache"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strconv"
+)
+
+func (rm *resourceManager) UpdateShardConfiguration(
+	ctx context.Context,
+	r *resource,
+	diffReporter *ackcompare.Reporter,
+) (*resource, error) {
+	input, err := rm.newUpdateShardConfigurationRequestPayload(r)
+	if err != nil {
+		return nil, err
+	}
+	resp, respErr := rm.sdkapi.ModifyReplicationGroupShardConfigurationWithContext(ctx, input)
+	if respErr != nil {
+		fmt.Printf("Failed to ModifyReplicationGroupShardConfigurationWithContext, Error: %v\n", respErr)
+		return nil, respErr
+	}
+	return provideUpdatedResource(r, resp.ReplicationGroup)
+}
+
+func (rm *resourceManager) UpdateReplicaCount(
+	ctx context.Context,
+	r *resource,
+	diffReporter *ackcompare.Reporter,
+) (*resource, error) {
+
+	for _, diff := range diffReporter.Differences {
+		if diff.Path == "Spec.ReplicasPerNodeGroup" {
+			desired, err1 := strconv.Atoi(diff.ValueA)
+			latest, err2 := strconv.Atoi(diff.ValueB)
+
+			if err1 != nil || err2 != nil {
+				return nil, fmt.Errorf("UpdateReplicaCount failed: invalid values")
+			}
+
+			if latest < desired { // increase
+				fmt.Printf("Requesting Increase Replica Count. Old value: %v, New Value: %v\n", latest, desired)
+				input, err := rm.newIncreaseReplicaCountRequestPayload(r)
+				if err != nil {
+					fmt.Printf("Error occurred: %v\n", err)
+					return nil, err
+				}
+				resp, respErr := rm.sdkapi.IncreaseReplicaCountWithContext(ctx, input)
+				if respErr != nil {
+					fmt.Printf("Failed to IncreaseReplicaCountWithContext. Error: %v\n", respErr)
+					return nil, respErr
+				}
+				return provideUpdatedResource(r, resp.ReplicationGroup)
+			} else { // decrease
+				input, err := rm.newDecreaseReplicaCountRequestPayload(r)
+				fmt.Printf("Requesting Decrease Replica Count. Old value: %v, New Value: %v\n", latest, desired)
```

Comment:
this is fine for now. Eventually we'll plumb a logger down into the resourceManager so you can use `rm.log.V(1).Info()` here for debug log statements.

Reply:
👍